### ISSUE TITLE
feature: implement advanced change detection

### ISSUE THREAD

**Issue by ivandeex** [rclone/rclone#4810]
**(Q)** Currently rclone provides `copy` and `sync` commands spinning around three basic traits (_hashsum_, _modtime_, _size_) to detect changes using internal [equal()](https://github.com/rclone/rclone/blob/v1.53.3/fs/operations/operations.go#L136) function and its helpers.
The latter used to grow _organically_ in response to the stream of [change detection](https://github.com/rclone/rclone/issues?q=label%3A%22change+detection%22) feature requests, often serving contradictory user requirements such as _consistency_ (this PR https://github.com/rclone/rclone/pull/4011#issuecomment-592186479), _cost savings_ (see https://github.com/rclone/rclone/issues/2889#issuecomment-508683225) and many others (sometimes [unexplained ;)](https://github.com/rclone/rclone/issues/4793)), consequently blurring the line between "bug" ang "feature".

This in turn has led to a number of change-related CLI flags (see https://rclone.org/flags), some of them having negative (_no-check_), conditional (_x-if-y_) or even double-negative (_noignore_) names, which interact in exotic ways with 30+ supported backends (https://rclone.org/overview#features) and their (in)ability to provide required traits.

On top of that we have [filtering](https://github.com/rclone/rclone/issues?q=label%3A%22filtering%22) flags and thin difference of `sync` vs `copy`. I feel that many users have problems telling it, and documentation doesn't stress enough.

In the end we might expect that _**any** behavior can be requested_, even that that feels esoteric or illogical.
¯\\\_(ツ)\_/¯

**(A)** This feature request is up to:
- inventory existing flags related to change detection
- assign ordering priorities (weights?) to flags and conditions
- describe which flag(s) win if they are used in combination
- probably develop a syntax similar (in idea) to Unix PAM (see https://linux.die.net/man/5/pam.conf or https://docs.ansible.com/ansible/latest/collections/community/general/pamd_module.html#examples) or NSS directives (https://linux.die.net/man/5/nsswitch.conf)
- (I started thinking about CSS rule weights but stopped myself ;))
- see whether existing flags can be described by the new syntax
- make it as simple as possible (think okkam's razor)
- revise test coverage, make sure existing shell wrappers around rclone don't break
- add a simple enough automata in _equal_ to handle the logic
- add a combined flag like `--change-criteria +size,-hash,!modtime,?whatever` with a syntax reach enough to cover existing and esoteric cases
- improve docs: add examples
- improve docs: map the combined flag on existing flags
- improve docs (other)

This seems like a big effort covering <5% of user base. OTOH there is a _generalization_ PR https://github.com/rclone/rclone/pull/3782 (multiwrite union) that has been a [success](https://github.com/rclone/rclone/pull/3782#issuecomment-601892715).

(spun off the https://github.com/rclone/rclone/pull/4011#issuecomment-734246794)

**Comment by ivandeex** [comment]
(from https://github.com/rclone/rclone/pull/4011)

> However it serves as a good example of uncontrolled complexity (see [#4793 (comment)](https://github.com/rclone/rclone/issues/4793#issuecomment-733647201)) of the ["equal" method](https://github.com/rclone/rclone/blob/v1.53.3/fs/operations/operations.go#L136) and around, which used to grow _organically_ in response to the stream of [change detection](https://github.com/rclone/rclone/issues?q=label%3A%22change+detection%22) feature requests, often serving contradictory user requirements such as _consistency_ (this PR [#4011 (comment)](https://github.com/rclone/rclone/pull/4011#issuecomment-592186479)), _cost savings_ (see [#2889 (comment)](https://github.com/rclone/rclone/issues/2889#issuecomment-508683225)) and many others (sometimes [unexplained ;)](https://github.com/rclone/rclone/issues/4793)), blurring the line between "bug" ang "feature".

I think we could inventory the flags to see where we are at.

However I don't want to break backwards compatibility, so if we want to change stuff then we need to do so carefull!

The `equal` function has been reasonably stable, I think the only thing added to it recently is --refresh-times - which is a very specialised options.

User demand is how rclone has gained new features, though we need to be careful to push back on ones which aren't of general interest.

> This in turn has led to a number of change-related CLI flags (see https://rclone.org/flags), some of them having negative (_no-check_), conditional (_x-if-y_) or even double-negative (_noignore_) names, which interact in exotic ways with 30+ supported backends (https://rclone.org/overview#features) and their (in)ability to provide required traits.

Indeed. N-flags * N-providers * N-options-per-provider is a big number.

> On top of that we have _filtering_ flags and two unobviously different commands: `sync` vs `copy`. I feel that many users won't feel difference between them and documentation doesn't stress enough.

More docs for `sync` and `copy` would be appreciated. The `sync` page has lots of warnings on now as a result of user accidents.

> In the end we might expect that _**any** behavior can be requested_, even that that feels esoteric or illogical.

Yes, users want to get their stuff done, it is our job to keep rclone consistent!
 
> This probably calls for some brushing ([#4793 (comment)](https://github.com/rclone/rclone/issues/4793#issuecomment-733628218)):

I'll just copy the comment here

> Idea: pause adding new --no-ignore-x-if-prefer-y flags for now. Add examples to docs for existing flags, they cover 95% of known use cases. For remaining 5%, create an issue to design a flag like --change-criteria +time,-size,whatever.

I think that is a good plan to control the complexity.

> * open a dedicated _issue_ with bullets below
> * assign ordering priorities (weights?) to flags and conditions
> * describe which flag wins if they are used in combination
> * probably develop a syntax similar (in idea) to Unix PAM (see https://linux.die.net/man/5/pam.conf or https://docs.ansible.com/ansible/latest/collections/community/general/pamd_module.html#examples) or NSS directives (https://linux.die.net/man/5/nsswitch.conf)
> * (I started thinking about CSS rule weights but stopped myself ;))
> * see whether existing flags can be described by the new syntax
> * make it as simple as possible (think okkam's razor)
> * revise test coverage, make sure existing shell wrappers around rclone don't break
> * add a simple enough automata in _equal_ to handle the logic
> * add a combined flag like `--change-criteria +size,-hash,!modtime,?whatever` with a syntax reach enough to cover existing and esoteric cases
> * improve docs: add examples
> * improve docs: map the combined flag on existing flags
> * improve docs (other)
> 
> This seems like a big effort covering <5% of user base so I'm not sure.

Making a generalised solution will mean that advanced users will have somewhere to go when then need to get stuff done, but hopefully a much smaller subset of flags will cover the main use cases (eg --checksum, --size-only).

I like that idea. We could implement it then deprecate flags which we don't like. (You can hide flags so they still work but are hidden).

> OTOH there is a _generalization_ PR #3782 (multiwrite union) that has been a [success](https://github.com/rclone/rclone/pull/3782#issuecomment-601892715).

Great idea.

I especially like your moratorium on adding more complexity to what we've got already.

Do you want to make an issue, then we can think about all the things which might need to be controlled?

**Comment by ivandeex** [comment]
Related to https://github.com/rclone/rclone/issues/4815 (separate "sync" page in docs)

**Comment by ivandeex** [comment]
Let me copy here few bits of the discussion about `non-commutive` flags and `rclonic` file comparison from https://github.com/rclone/rclone/pull/5164#issuecomment-842407758 and https://github.com/rclone/rclone/pull/5164#issuecomment-843151817 to prevent burying it.

### Filters

Why are the filters different? The only one of concern is `--exclude-if-present` which can cause issues with bi-directional sync. The following is the note I include in my readme about them:

> Filters work well since they are applied to both sides. This means that changing a filter will make the files look like it was deleted on *both* sides and nothing will happen.
> 
> However, the [`--exclude-if-present`][eifp] filter is **very dangerous** if the excluded file is added on one side only *after* files in the directory are in sync. It will cause rclone to skip the directory on one side and appear like it was deleted. 
> 
> There *are* safe ways to use `--exclude-if-present`. For example, you can place exclude file in place on *both* sides before syncing. (Or placing it without adding the filter, syncing, and then adding the filter). Or, if the files have *never* been synced before (i.e. it's a new directory with the exclude file) then nothing bad will happen.

This isn't a new problem in bisync from regular rclone though I do see it amplified for it. Again this goes back to the philosophical idea of a sync-config vs CLI flags. The reason I chose not to use the CLI flags was because I wanted it very clear that a "sync job" is not a one-off.

### --exclude-if-present

> > the [`--exclude-if-present`] filter is **very dangerous** if the excluded file is added on one side only _after_ files in the directory are in sync. It will cause rclone to skip the directory on one side and appear like it was deleted.

We have open bug reports (#4523) about this.
We have requests (#4753, #4793) to add more _non-commutive_ flags. 
We try (#4011, #4856) to restrain reasonably.

### `rclonic` file comparison

> One of the many features of rclone is that it can handle nearly all  remotes the same with minor details.
> I really think it would have been better to *not* base this approach on an existing tool and instead design the algorithm to be "rclonic"

The heart of rclone's omnivory change detection (runtime) lies around [operations.equal](https://github.com/rclone/rclone/blob/v1.55.1/fs/operations/operations.go#L167), [operations.checkHashes](https://github.com/rclone/rclone/blob/v1.55.1/fs/operations/operations.go#L67), et cetera. It's kept afloat by the astute test framework, in particular [fstest.CheckListing](https://github.com/rclone/rclone/blob/v1.55.1/fstest/fstest.go#L264), [fstest.CheckHashes](https://github.com/rclone/rclone/blob/v1.55.1/fstest/fstest.go#L137), [fstest.CheckTime](https://github.com/rclone/rclone/blob/v1.55.1/fstest/fstest.go#L116) and [<i>drums</i>] [retrying](https://github.com/rclone/rclone/blob/v1.55.1/fstest/test_all/run.go#L400).

This code is not big in terms of LOC but its [perceived complexity](https://github.com/rclone/rclone/issues/4793#issuecomment-733647201) is high... while users continue to request for more [change detection](https://github.com/rclone/rclone/issues?q=is%3Aopen+is%3Aissue+label%3A%22change+detection%22) and [filtering](https://github.com/rclone/rclone/issues?q=is%3Aopen+is%3Aissue+label%3Afiltering) features.
My first approach to bisync about three months ago came across problems I can't clearly remember now so I tried it differently.

There is an ongoing proposal (#4810) that seeks to put this complexity under control. Ultimately we will need to inventory the state of affairs if we want to continue adding more features safely. I think we could continue this discussion in #4810.

### Standing on the shoulders of `rsync`

> when uncertain how to handle a situation, look at rsync.

Unfortunately rclone has already diverged (#1228, #3375, #5040, #5043, [docs](https://rclone.org/filtering/#how-filter-rules-are-applied-to-directories)) from rsync
We can't easily fix that due to backwards compatibility concerns so there exists a waiting proposal (#4776) of `rclone rsync` besides `rclone sync`.

PS @ncw @jwink @ivandeex

**Comment by ivandeex** [comment]
(https://forum.rclone.org/t/syncing-timestamps-when-using-checksum/24744/12)
Yet another mode unsupported by current set of flags:

> My use case is syncing files from source to destination. After syncing I want all files on the destination have the same contents and modification times as the source.

> In most cases, the default sync functionality will achieve this effect. However, this does not actually verify checksums for files with matching timestamps, so there's a possibility that some files will be missed.

**Comment by ivandeex** [comment]
Related feature request:

https://forum.rclone.org/t/copy-overwrite-rules-size-rule/26807

> overwrite only if the destination file is smaller than the source file.

**Comment by ivandeex** [comment]
- #5821 

> if the backend doesn't deliver a proper hash for the file, rclone should rather upload the file, assuming there is a difference, than assuming no difference

**Comment by ncw** [comment]
I'm beginning to think that if we want to support all the things users want to do with syncing we are going to need to provide a scriptable sync engine... So imagine a (insert scripting language) fragment which dealt with the sync flow.

**Comment by ivandeex** [comment]
Still feels vague to me.
Will understand when we collect enough cases, allocate time and start architecting the change detection engine.

**Comment by ivandeex** [comment]
Still feels vague to me.
Will understand when we collect enough cases, allocate time and start architecting the change detection engine.
The 90% cases are already covered by existing flags.
When we architect the engine for remaining 10%, we a free to decide that we prefer
some NSS-like notation backed by a state machine that covers 90% of remaining 10% rather than embedding a LUA-like script.
The unlucky 1% (10% of 10%) i.e. 1% will be left for users to script around CLI/API. Excuse? rclone is not Santa Claus! :)

**Comment by ivandeex** [comment]
- https://github.com/rclone/rclone/pull/5576

> I use rclone mount with a google drive backend.
> When I  alter a file but leave data the same, a new revision will be generated without the patch

advanced change engine is expected to support behavior like the patch below
https://github.com/rclone/rclone/blob/34c28b29b690b16185c33f9188c08c9e3209b023/vfs/vfscache/item.go#L584-L592

**Comment by ivandeex** [comment]
- #4793
- #4856

**Comment by ivandeex** [comment]
- #4753

> We have a system where we need to copy files from an SFTP server to a Google Cloud Storage bucket. Our customers upload files into subdirectories on the SFTP server and when they're done, they upload an empty file named foobar_{timestmap}.xml. We wish to transfer the entire sub-directory from the SFTP server to GCS only once this file has been uploaded.
